File: bachelors/year3/semestre1/numerical_computational_methods/Lab 3/code.cpp
#define _CRT_SECURE_NO_WARNINGS
#include <iostream>
#include <algorithm>
#include <vector>
#include <Windows.h>
#include <initializer_list>
#include <iomanip>
#include <valarray>

using namespace std;

typedef valarray<double> args;

typedef double (*functype)(args);

class EqSystem {
protected:
    vector<pair<functype, functype>> syst;
public:
    EqSystem(initializer_list<pair<functype, functype>> lst) : syst(lst) {}
    args process(args a) {
        args res(a.size());
        for (int i = 0; i < res.size(); ++i)
            res[i] = (syst[i].second)(a);
        return res;
    }
    double get(int idx, args a) {
        return (syst[idx].first)(a);
    }
    args get(args a) {
        args res(a.size());
        for (int i = 0; i < res.size(); ++i)
            res[i] = (syst[i].first)(a);
        return res;
    }
};

double step = 0.01;

auto sqr = [](double x) {
    return x * x;
};

auto Abs = [](double x) {
    return x < 0 ? -x : x;
};

double norm(args a) {
    double res = 0;
    for (auto i : a)
        res += abs(i);
    return res;
}

double eps = 1e-7;

args approx(args a0, EqSystem st) {
    for (;;) {
        args cur = st.process(a0);
        if (norm(a0 - cur) < eps)
            return cur;
        a0 = cur;
    }
}

void output(args res) {
    for (auto i : res)
        cout << i << " ";
    cout << endl;
}


auto F0 = [](args a) { return sqr(a[0]) / 4 + sqr(a[1]) + sqr(a[2]) / 9 - 1; };
auto F1 = [](args a) { return sqr(a[0]) + sqr(a[1]) / 9 + sqr(a[2]) / 4 - 1; };

auto phi0 = [](args a) { return 2 * sqrt(1 - sqr(a[2]) / 9 - sqr(a[1])); };
auto phi1 = [](args a) { return 3 * sqrt(1 - sqr(a[2]) / 4 - sqr(a[0])); };
auto phi2 = [](args a) { return a[2]; };

double c = 0.3;
auto X = [](double z0) { return sqrt((1 - sqr(z0) / 4) / (1 + 9 / sqr(c))); };
auto Y = [](double z0) { return sqrt((1 - sqr(z0) / 9) / (1 + 4 / sqr(c))); };

int main() {
    //freopen("output.txt", "w", stdout);

    /*EqSystem st = {
        { [](args a) { return 5 * a[0] - 6 * a[1] + 20 * log(a[0]) + 16; }, [](args a) { return exp((6 * a[1] - 6 * a[0] - 16) / 20); } },
        { [](args a) { return 2 * a[0] + a[1] - 10 * log(a[1]) - 4; }, [](args a) { return exp((2 * a[0] + a[1] - 4) / 10); } }
    };

    args res = approx({ 1.25, 0 }, st);
    output(res);
    output(st.get(res));*/

    /*EqSystem st = {
        { [](args a) { return 3 * sqr(a[0]) + 1.5 * sqr(a[1]) + sqr(a[2]) - 5; }, [](args a) { return sqrt(5 - 1.5 * sqr(a[1]) - sqr(a[2])) / 3; } },
        { [](args a) { return 6 * a[0] * a[1] * a[2] - a[0] + 5 * a[1] + 3 * a[2]; }, [](args a) { return (6 * a[0] * a[1] * a[2] - a[0] + 3 * a[2]) / (-5); } },
        { [](args a) { return 6 * a[0] * a[2] - a[1] * a[2] - 1; }, [](args a) { return (a[1] * a[2] + 1) / (6 * a[0]); } }
    };

    args res = approx({1.25, 0, 0.25}, st);
    output(res);
    output(st.get(res));*/

    /*EqSystem st = {
        { [](args a) { return 2 * sqr(a[0]) - a[0] * a[1] - 5 * a[0] + 1; }, [](args a) { return sqrt((a[0] * (a[1] + 5) - 1) / 2); } },
        { [](args a) { return a[0] + 3 * log(a[0]) / log(10) - sqr(a[1]); }, [](args a) { return sqrt(a[0] + 3 * log(a[0]) / log(10)); } }
    };

    args res = approx({ 3.5, 2.5 }, st);
    cout << res[0] << " " << res[1] << endl;*/

    /*EqSystem st = {
        {F0, phi0},
        {F1, phi1},
        {nullptr, phi2}
    };

    for (double zi = 0; zi <= 0; zi += step) {
        args s = approx({ X(zi), Y(zi), zi }, st);
        cout << s[0] << " " << s[1] << " " << s[2] << " " << st.get(0, s) << " " << st.get(1, s) << endl;
    }*/

    system("pause");
    return 0;
}

File: bachelors/year3/semestre1/numerical_computational_methods/Lab 3/code.py
'''import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
from math import *
from numpy import *

step = 0.01
Z = arange(-2, 2, step)
X, Y = [], []
fig = plt.figure()
ax = fig.add_subplot(111, projection='3d')


def F0(x, y, z):
    return (x**2) / 4 + (y**2) + (z**2) / 9 - 1
def F1(x, y, z):
    return (x**2) + (y**2) / 9 + (z**2) / 4 - 1
for p in [0, 1, 2, 3]:
    c1 = 1 if p % 2 == 0 else -1
    c2 = 1 if (p / 2) % 2 == 0 else -1
    X, Y = [], []
    for zi in Z:
        xi, yi = c1 * sqrt(288 - 77 * (zi**2)) / (3 * sqrt(35)), c2 * sqrt(108 - 7 * (zi**2)) / (2 * sqrt(35))
        X.append(xi)
        Y.append(yi)
        print xi, yi, zi, F0(xi, yi, zi), F1(xi, yi, zi)
    ax.plot(X, Y, Z)    


plt.show()
Axes3D.plot()'''
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
from math import *
from numpy import *

def sqr(a):
    return a * a

def F0(x, y, z):
    return (x**2) / 4 + (y**2) + (z**2) / 9 - 1
def F1(x, y, z):
    return (x**2) + (y**2) / 9 + (z**2) / 4 - 1

def norm(x, y):
    return abs(x) + abs(y)

def f0(x, y, z):
    return (x**2) / 4 + (y**2) - (1 - z * z / 9) * (norm(x, y) ** 2)
def f1(x, y, z):
    return (x**2) + (y**2) / 9 - (1 - z * z / 4) * (norm(x, y) ** 2)

T = 0.2

def phi0(x, y, z):
    #return x - T * f0(x, y, z)
    return 2 * sqrt(fabs(1 - sqr(z) / 9 - sqr(y)))
def phi1(x, y, z):
    #return y - T * f1(x, y, z)
    return 3 * sqrt(fabs(1 - sqr(z) / 4 - sqr(x)))

eps = 0.001

def approx(x0, y0, z0):
    while True:
        x = phi0(x0, y0, z0)
        y = phi1(x0, y0, z0)
        if (norm(x - x0, y - y0) < eps):
            return [x, y]
        x0, y0 = x, y

step = 0.1
Z = arange(0, 1 + step, step)
X, Y = [], []
fig = plt.figure()
ax = fig.add_subplot(111, projection='3d')

c = 0.3
def X0(z):
    return sqrt((1 - sqr(z) / 4) / (1 + 9 / sqr(c)))
def Y0(z):
    return sqrt((1 - sqr(z) / 9) / (1 + 4 / sqr(c)))

for zi in Z:
    xi, yi = approx(X0(zi), Y0(zi), zi)
    X.append(xi)
    Y.append(yi)
    print xi, yi, zi, F0(xi, yi, zi), F1(xi, yi, zi)

ax.plot(X, Y, Z)
plt.show()
Axes3D.plot()